experanto/data.py

```diff
@@ -74,8 +101,141 @@ def __getitem__(self, idx):
+        )
+        self.device_names = self._experiment.device_names
+        # this is needed to match sensorium order only
+        if "screen" in self.device_names and "responses" in self.device_names:
+            start = ["screen", "responses"]
+            self.device_names = tuple(start) + tuple(
+                set(self.device_names).difference(set(start))
+            )
+        self.subsample = subsample
+        self.cut = cut
+        self.add_channel = add_channel
+        self.channel_pos = channel_pos
+        self.rescale = rescale
+        assert (
+            0 <= channel_pos < 4
+        ), "channels could be extended only for positions [0,3]"
+
+        self.start_time, self.end_time = self._experiment.get_valid_range("screen")
+        self.DataPoint = namedtuple("DataPoint", self.device_names)
+        self.MetaNeuro = namedtuple(
+            "MetaNeuro", ["cell_motor_coordinates", "unit_ids", "fields"]
+        )
+        self._read_trials()
+
+    def _read_trials(self):
+        # reads all videos from valid tiers and saves times for them
+        # also have saves the start and end time if test videos are in between
+        # todo
+        screen = self._experiment.devices["screen"]
+        self._trials = [
+            t
+            for t in screen.trials
+            if isinstance(t, VideoTrial) and t.get_meta("tier") == self.tier
+        ]
+        s_idx = np.array([t.first_frame_idx for t in self._trials])
+        # todo - not sure if it should be t.first_frame_idx + t.num_frames
+        e_idx = np.array([t.first_frame_idx + t.num_frames - 1 for t in self._trials])
+        # todo - this uses the assumption that sampling_rate is less or equal the sampling rate of the screen stimuli
+        if self.cut:
+            assert all(
+                [t.num_frames >= self.stim_duration for t in self._trials]
+            ), "stim_duration should be smaller"
+        if len(s_idx):
+            self._start_times = screen.timestamps[s_idx]
+            self._end_times = screen.timestamps[e_idx]
+        else:
+            self._start_times = np.array([])
+            self._end_times = np.array([])
+
+    def __len__(self):
+        return len(self._trials)
+
+    @property
+    def neurons(self):
+        loc_meta = {
+            "cell_motor_coordinates": self._experiment.devices[
+                "responses"
+            ].cell_motor_coordinates,
+            "unit_ids": self._experiment.devices["responses"].unit_ids,
+            "fields": self._experiment.devices["responses"].fields,
+        }
+        return self.MetaNeuro(**loc_meta)
+
+    def __getitem__(self, idx):
+        """
+
+        :param idx: idx of the video
+        :return: this would return video in data['screen'] with shape of [t, h, w]
+        """
+        fs = self.sampling_rate
+        times = np.arange(self._start_times[idx], self._end_times[idx], 1 / fs)
+        # get all times possible
+        # cut is needed
+        if self.cut:
```

Comment:
I dont fully understand this here, but maybe we can chat about this

Reply:
this is mostly for sensorium backcompatibility tests 
there we subsample like N frames from each video during training
but during validation we take the whole video
so if cut is false - it would return the whole video, like during validation

Reply:
Ok I understand! I see that it makes sense to have a dataset that reproduces sensorium exactly.

In the general case, for a general purpose video dataset, this could be a also added to the list of transforms. but let's do that after we come up with a good principle